tree/treeplayer/src/TTreeReaderArray.cxx

```diff
@@ -532,7 +532,25 @@ void ROOT::Internal::TTreeReaderArrayBase::CreateProxy()
          return;
       }
 
-      if (fDict != branchActualType) {
+      auto matchingDataType = [](TDictionary *left, TDictionary *right) -> bool {
+         if (left != right) {
+            auto left_datatype = dynamic_cast<TDataType *>(left);
+            auto right_datatype = dynamic_cast<TDataType *>(right);
```

Comment:
Can you return false if one of the dynamic_casts fails, otherwise extract GetType() into local variables so you don't need to call that 12 times below?

Reply:
Done.